### ISSUE TITLE
How to "wait" until a "dependency loop" finishes?

### ISSUE THREAD

**Issue by ezyang** [ndmitchell/shake#454]
The motivating example is hs-boot and Template Haskell, see this bug: https://ghc.haskell.org/trac/ghc/ticket/1012 . Consider the following files:

* `A.hs-boot`
* `B.hs` imports `A.hs-boot`
* `A.hs`
* `C.hs` imports `B.hs`, but NOT `A.hs`. Furthermore, it uses a Template Haskell splice defined in `B.hs`

In this case, in order to run the Template Haskell splice in `C.hs`, we need to have running executable code for `B.hs`. However, in general, we don't actually have a runnable code for `B` until we build `A`.

The obvious Shake-ification of these build rules will fail to have an implicit dependency for `C.hs` on `A.hs`. So how can we arrange for this dependency to happen? The best I can think of is to create a set of dummy targets representing whether or not a module is executable or not (note, you CAN'T use presence of .o files, since we do generate that when we build B.hs, it's just not runnable because there is no A.o). The dependency structure then is that runnable:A.hs-boot depends on A.o and the runnable deps of its dependencies, runnable:B.hs depends on the runnable deps of its dependencies, i.e. runnable:A.hs-boot a.k.a. A.o, and A.hs depends on the runnable deps of its dependencies, but NOT itself (there is no way to depend on targets to get this information: you'd have to get the lists of runnable deps and then merge them, filtering out A.)

Is there a better way?

**Comment by Mathnerd314** [comment]
I believe the code is as follows, modified from [that ticket](https://ghc.haskell.org/trac/ghc/ticket/1012#comment:5) (swapping B and C)
```
ModuleA.hs-boot:
module ModuleA where

ModuleA.hs:
module ModuleA where
import ModuleB

ModuleB.hs:
module ModuleB where
import Language.Haskell.TH
import {-# SOURCE #-} ModuleA
nothing = return [] :: Q [Dec]

ModuleC.hs:
{-# LANGUAGE TemplateHaskell #-}
module ModuleC where
import ModuleB
$(nothing)
```

This successfully compiles with this sequence (and omitting A.hs makes it fail):
```
  ghc -c -dynamic ModuleA.hs-boot -- generates A.o-boot
  ghc -c -dynamic ModuleB.hs      -- generates B.o, consumes A.o-boot
  ghc -c -dynamic ModuleA.hs      -- generates A.o, consumes B.o
  ghc -c -dynamic ModuleC.hs      -- generates C.o, consumes A.o B.o
```

Simon Marlow has a solution to get the extra A dependency:
> Every module that depends on a module in a cycle, but is not a member of that cycle, should have an implicit dependency on each of the modules in the cycle. 

This translates to these dependencies (tracking .o or .hi files is a separate problem, after #453):
```
A.hs -> B.hs
B.hs -> A.hs-boot, mark as cycle
C.hs -> B.hs, cycle of B.hs (detected)
cycle of B.hs -> A.hs B.hs
```
The cycle constraint can probably be relaxed somehow so that the cycle rule doesn't have to actually detect cycles, but regardless, it seems reasonable.

**Comment by ezyang** [comment]
Yeah. So essentially, there's a natural dependency structure you get by looking at the cycle, but there isn't really a good way to talk about cycles in Shake.
